lib/services/publish-service.ts

```diff
@@ -0,0 +1,208 @@
+﻿///<reference path="../.d.ts"/>
+"use strict";
+
+import options = require("../common/options");
+import os = require("os");
+import path = require("path");
+import util = require("util");
+var Table = require("cli-table");
+
+export class PublishService implements IPublishService {
+    private static JSON_PUBLISH_FILE_NAME = ".abpublish";
+    private static ALLOWED_CONNECTION_TYPE = "ftp";
+    private static ERROR_MESSAGE_WHEN_PROJECT_EXISTS_ON_REMOTE = "User cancelled operation";
+    private static PUBLIC_URL_DEFAULT_VALUE = "";
+    private static REMOVE_ALL_CONNECTIONS_MESSAGE = "All of them";
+
+    public allPublishConnections: IPublishConnection[];
+    public ftpPublishConnections: IPublishConnection[];
+
+    constructor(private $errors: IErrors,
+                private $fs: IFileSystem,
+                private $logger: ILogger,
+                private $progressIndicator: IProgressIndicator,
+                private $projectConstants: Project.IProjectConstants,
+                private $projectFilesManager: Project.IProjectFilesManager,
+                private $project: Project.IProject,
+                private $prompter: IPrompter,
+                private $server: Server.IServer) {
+
+        this.$project.ensureProject();
+
+        if (!this.$project.capabilities.publish) {
+            this.$errors.failWithoutHelp("The publish command is only applicable to projects of type %s", this.$projectConstants.TARGET_FRAMEWORK_IDENTIFIERS.MobileWebsite);
+        }
+
+        this.readConnections().wait();
```

Comment:
this can throw ($errors.fail actually throws). Because this is a constructor, yok would barf its unhappiness upon the user in the form of "Cannot construct blah-blah" message. Is this OK with you?